### PR TITLE
RFC: Motion definition & APIs

### PR DESCRIPTION
**Pull request by layershifter**
## Description

Request for Comments (RFC) proposes the use of the Web Animations API for defining and utilizing motion animations in Fluent UI React components. It discusses various aspects including motion definition, APIs, usage & replacement, performance, and potential issues, while also comparing it with the CSS-based approach.

----

[Preview link 🔗](https://github.com/layershifter/office-ui-fabric-react/blob/docs/motion-rfc/rfcs/react-components/convergence/motion-definition-n-apis.md)

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/62312d15851d9ec28b2e84216e2ba2aa19b14d70'>62312d15851d9ec28b2e84216e2ba2aa19b14d70</a></sub>
<div id="bundleSizeComment9423"/>

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 1]
```diff
@@ -0,0 +1,999 @@
...
+As with `createAtom()`, the factory returns a React component that clones the child element and applies the provided animation to it:
+
+```tsx
+import { createPresence, presence } from '@fluentui/react-motions-preview';
+
+const FadeSlow = createPresence(presence.fade.slow());
+
+function MyComponent() {
+  return (
+    <FadeSlow>
```
Comment (ling1726) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:449]: should the `visible` prop be used in the example? (and others like appear/unmountOnExit)

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 2]
```diff
@@ -0,0 +1,999 @@
...
+</details>
+
+### Using presence motions
+
+Compared to the CSS option, we don't need to apply classes to the element, so factories don't need to control state and won't suffer from the same issues as the CSS option with re-rendering.
+
+> The Web Animations API has its own state machine [by design](https://developer.mozilla.org/en-US/docs/Web/API/Animation/playState), so we can subscribe, for example, to the onfinish event to handle unmount.
+
+<details>
+<summary>Handling unmount using callbacks</summary>
```
Comment (ling1726) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:413]: It's not very from this example how the animation events relate to the presence components that are created by this API

I think there should be some way to link the unmount handling to `createPresence` since that is essentially how a presence handles unmounting on exit.

Perhaps you could reduce both details blocks to one single details block?

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 3]
```diff
@@ -0,0 +1,999 @@
...
+```tsx
+const fadeEnterSlow: MotionAtom = {
+  keyframes: {
+    from: { opacity: 0 },
+    to: { opacity: 1 },
+  },
+  fill: 'forwards',
+
+  // Heads up! `duration` is in milliseconds and `easing` is a string
+  // ⚠️ We can't use CSS variables there
```
Comment (mnajdova) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:301]: **Disclaimer**: I am not leaving this comment as a recommendation for what is best for MSFT, but talking more in general terms 😄 

With moving to CSS variables + static styles extraction, I expect more and more theme definitions to depend on CSS variables - including here the tokens related to animations. Also, with the rose of RSC I would suppose that you would want to limit the requirement for using context (unless you are opting for the server context) to a minimum (in case you want these motion tokens to be defined via some kind of context). This is my only worry with this kind of API.

Tagging @brijeshb42 and @siriwatknp in case they want to give their thoughts on this RFC.

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 4]
```diff
@@ -0,0 +1,999 @@
...
+## Summary
+
+This RFC proposes the way to define motion animations and APIs to use them in Fluent UI React components (`@fluentui/react-components`).
+
+## Background
+
+There are three main use cases that this RFC is trying to solve:
+
+- **Motion definition**: How should motion animations be defined?
+- **Motion APIs**: Which APIs should be exposed for utilizing motions in Fluent UI React components?
```
Comment (mnajdova) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:59]: Do you plan to restrict developers to only be able to override motions using this API? Can users override the default motions using plain CSS for e.g.?

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 1 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md]: Good catch, indeed. Updated 98e43f5

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 2 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md]: I tried to improve this in c4a74dd. Does it make sense now or should I just remove this block?

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 3 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:332]: @mnajdova that's a good callout. The original explanation was not complete and confusing 😕 I added a new section that explains CSS vars usage in 6dd795c 👍 

TL;DR CSS variables could be used in [`keyframes`](https://react.fluentui.dev/?path=/docs/utilities-web-motions-preview--motion-default#tokens-usage) as is. The limitation appears only for [`options`](https://developer.mozilla.org/en-US/docs/Web/API/KeyframeEffect/KeyframeEffect#options), but as motions could be defined as factories - there is a workaround for it.

Comment (mnajdova) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:332]: > The limitation appears only for options, but as motions could be defined as factories - there is a workaround for it.

Thanks for the explanation, it's good to know that it can be solved in a way. However, I suppose this means that the component would need to re-render for the change to be applied no? You would have to force re-render if a CSS var changes, while with CSS animations that won't be the case. Anyway, this feature may not be in the requirements you have, so maybe it doesn't matter that much.

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 5]
```diff
@@ -0,0 +1,1079 @@
...
+    }, []);
+
+    return React.cloneElement(props.children, { ref: elementRef });
+  });
+}
+```
+
+</details>
+
+### Motion definition
```
Comment (ling1726) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:285]: I think this RFC also needs to cover reduced motion use cases across the different proposed solutions.

For `useMotion` and RTG, the answer is fairly simple since CSS media queries could be applied to override the default animation styles.

For the WAPI motion definitions, I currently don't see a clear way to handle reduced motion from this proposal. I guess there should be some kind of matchMedia logic going on, but there could be multiple ways to integrate into the factory APIs, please show how that would be possible.

Comment (robertpenner) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:285]: There is a reduced motion implementation for perusal, pending an RFC writeup:

https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-motions-preview/src/hooks/useIsReducedMotion.ts

https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-motions-preview/src/factories/createAtom.ts#L47

Comment (robertpenner) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:285]: I would like to have a "developer toggle" for enabling reduced motion directly in JavaScript, for testing purposes. This would allow us to do things like show the same component with and without reduced motion side-by-side in Storybook. Whereas if reduced motion is only switchable globally, via the browser, we could not make this comparison simultaneously in the same page.

Comment (ling1726) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:285]: It's also important to note that reduced motion does not necessarily mean that the duration of the animation is reduced. In an animation of an element where its size and color changes, the color change can still be desirable

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 6]
```diff
@@ -0,0 +1,1079 @@
...
+}
+```
+
+This approach does not have obvious pros or cons, but it's worth noting that every animated component will have to be composed, for example:
+
+```tsx
+// ⚠️ This is not proposed API, it's just an example
+
+import { Button, Image } from '@fluentui/react-components';
+import { createAtom, createAtomElement, atom } from '@fluentui/react-motions-preview';
```
Comment (marcosmoura) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:430]: This is a neat API for creating reusable animation components based on new or existing atoms, but the naming here might conflict with well-established state/store libraries in the React ecosystem, like Recoil/Jotai, and possibly other libraries that uses the "atom" concept. 

Working with the concept of atoms does not necessarily mean that we need to name them "atom" and "createAtom".

If a project uses two libraries that implement "atoms", it might be confusing:

``` ts
import { createAtom, atom: motionAtom } from '@fluentui/react-motions-preview';
import { atom } from 'recoil'

// ... code

// Both are "global" scoped
const FadeEnter = createAtom(motionAtom.fade.enterSlow);
const todoListState = atom({
  key: 'TodoList',
  default: [],
});

// ... code
```

Even though the import comes from `@fluentui/react-motions` or `@fluentui/react-motion` and they can be renamed to something else, it might still cause confusion.

Could we evaluate if it is worth renaming it to something else, more scoped?

Something like: 

``` ts
import { createMotion, motion } from '@fluentui/react-motions-preview';

const FadeEnter = createMotion(motion.fade.enterSlow);
```

Or even:

``` ts
import { createMotionAtom, motionAtom } from '@fluentui/react-motions-preview';

const FadeEnter = createMotionAtom(motionAtom.fade.enterSlow);
```

File: rfcs/react-components/convergence/motion-definition-n-apis.md [hunk 7]
```diff
@@ -0,0 +1,1079 @@
...
+
+const flash: MotionAtom = {
+  keyframes: [
+    { backgroundColor: 'white' },
+    // 💡`tokens.colorBrandBackground` is a CSS variable i.e. `var(--colorBrandBackground)`
+    { backgroundColor: tokens.colorBrandBackground },
+    { backgroundColor: 'white' },
+  ],
+
+  duration: 500,
```
Comment (marcosmoura) [inline rfcs/react-components/convergence/motion-definition-n-apis.md:323]: Can we think of a way to specify `duration` and `easing` for each stage of an animation?

One example is the Drawer and Dialog components, which have different easing curves for each stage (entering or exiting). 
For presence-like animations, it is also quite useful to have "entering" faster than "exiting" animations.

**Comment by marcosmoura** [comment]
Leaving this comment globally here, as there were no mentions about it, and I was unsure where it belongs within this document (or if it should be in this document or in another RFC).

The reason to create a broader solution to handling animations is that `useMotion` affects the way we re-render components and how to apply animations using it. But the useful part of the hook is to detect animations, no matter how they are applied.

Even though it is an advanced usage, there is a need to detect the state of an animation, with the most common one being for "presence" cases. That was the main reason to create the `useMotion` hook in the first place. 

Some examples:
- Detecting when something finished appearing or disappearing from the screen to display/hide another UI element. 
- Creating some sort of orchestrated animations, not based on a group of items (or staggered animations) but for separated/isolated UI elements.

Both examples can be seen on this [Code Sandbox](https://codesandbox.io/p/sandbox/stoic-tristan-9j3zrq), made based on a real-life scenario used by Loop application.

Can we think of a way to accommodate that scenario with the new proposed API?

In these cases, performance is important but might not be the top priority as it is an advanced case.

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 6 above)
Comment (marcosmoura) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:430]: I just realized this comment was misplaced, as it was not a proposed API but an example. But it is relevant for the whole API naming convention, so I'll keep it here.

**Comment by miroslavstastny** [comment]
@ling1726 to review on behalf of teams-prg

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 3 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:332]: > However, I suppose this means that the component would need to re-render for the change to be applied no? You would have to force re-render if a CSS var changes, while with CSS animations that won't be the case.

@mnajdova yeah, you're correct about this. However, even when a value changes for a variable it will not trigger a new animation 🤔

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 5 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:285]: @ling1726 I added a section in 07139a0 that explains current state and possible solution.

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 6 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:430]: I think that it's a good point. What about the following?

- `createAtom()` => `createMotionAtom()`
- `createPresence()` => `createMotionPresence()`

/cc @robertpenner

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 7 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:324]: @marcosmoura that's already handled by design. `MotionAtom` is an atomic unit while, `PresenceMotion` is a compound from `enter` and `exit` atoms. So the described case, we can do:

```ts
const dialogPresence: PresenceMotion = {
  enter: {
    keyframes: [],
    duration: 500,
  },
  exit: {
    keyframes: [],
    duration: 300,
  }
}
```

However, this made me realize that `PresenceMotion` type is not explained in the RFC, added a note in c79b772.

**Comment by layershifter** [comment]
> The reason to create a broader solution to handling animations is that `useMotion` affects the way we re-render components and how to apply animations using it. But the useful part of the hook is to detect animations, no matter how they are applied.
> 
> Even though it is an advanced usage, there is a need to detect the state of an animation, with the most common one being for "presence" cases. That was the main reason to create the `useMotion` hook in the first place.
> 
> Some examples:
> 
> * Detecting when something finished appearing or disappearing from the screen to display/hide another UI element.
> * Creating some sort of orchestrated animations, not based on a group of items (or staggered animations) but for separated/isolated UI elements.
> 
> Both mentioned examples can be seen on these examples: [Sync animation between components](https://stackblitz.com/edit/stackblitz-starters-y2kwu6?file=src%2FExample.tsx) [Detect animation state](https://stackblitz.com/edit/stackblitz-starters-xsa3tf?file=src%2FDrawerExample2.tsx) These examples are made based on a real-life scenario used by some apps, such as Loop.
>
> Please note that animationend/animationstart events can be used for this, but in case an element has multiple animations, the event will be triggered multiple times and with that be hard to detect when animations really finished.

@marcosmoura as we discussed offline, I'm not fully understand the significance of using a state machine in this context. Let me dig into both scenarios we discussed.

#### Sync animation between components

From a React perspective, `useMotion()` operates declaratively, responding to state updates:

```ts
const motion = useMotion(stateValue);
```

The example provided initiates three animations simultaneously in a `Drawer`, content, and footer. Achieving the same outcome is possible using the Web Animations API or any alternative approach, as the animations' trigger is the `stateValue`. For example:

```ts
function App() {
  return (
    <>
      <Drawer open={stateValue} />
      <Content visible={stateValue} />
      <Footer visible={stateValue} />
    </>
  );
}
```

### Detect animation state

As you mentioned, both examples could function without `useMotion()` since they revolve around animation states, which could be managed using animation callbacks. For example:

```tsx
function App() {
  const [open, setOpen] = React.useState(false);
  const [isUnmounted, setIsUnmounted] = React.useState(!open);

  return (
    <>
      <InlineDrawer
        onOpenChange={(_, data) => {
          if (data.open) {
            setIsUnmounted(true);
          }
        }}
        onTransitionEnd={() => {
          /* this is equal to `motion.type === 'unmounted'` */
          if (!open) {
            setIsUnmounted(false);
          }
        }}
      />
      <div className={mounted ? "x" : "y"} />
    </>
  );
}
```

Similar handling could be achieved for `motion.type === 'idle'`. I understand your concern about potential issues with animation events, so we can expose callbacks that emulate state, like so:

```tsx
<Collapse onMotionFinish={() => {}} />
```

> Also, those events do not take into consideration the animation-delay, so animationstart only triggers after the delay, so that is not ideal to detect the full duration of the animation.

Regarding your point about `animation-delay`... It doesn't seem problematic since we're already triggering motions based on state i.e. `stateValue`.

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 6 above)
Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md]: @marcosmoura I updated to RFC to use `createAtomMotion()` & `createPresenceMotion()`

File: rfcs/react-components/convergence/motion-definition-n-apis.md (hunk 7 above)
Comment (marcosmoura) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:328]: Maybe I was not completely clear here.
Elements can have multiple animations with different duration and/or easing configurations:
`animation: fade 1s ease-out, scale 2s ease-in`.

Also, an element can have different multiple animations depending on whether is opening/closing, for example:

``` css
.open {
	animation: open-fade 1s ease-in, open-scale 2s ease-in-out;
}

.close {
	animation: close-fade 1.5s ease-out, close-scale 2.5s ease-out;
}
```

How can we replicate that using the proposed API?


I can propose a way that would be the `createAtom` also accepting an array of motion definitions, instead of only being able to receive a single definition. A complex scenario would be:

``` ts
const fade: MotionAtom = {
  keyframes: [{ opacity: 0 }, { opacity: 1 }],
  duration: 500,
}

const scale: MotionAtom = {
  keyframes: [
    { transform: "scale(0)" }, 
    { transform: "scale(0.7)", offset: 0.5 }, 
    { transform: "scale(1)" }
  ],
  duration: 800,
}

const FadeScale = createAtom([scale, fade])
```

Internally, it would call the `animate` method multiple times, for each array item.

Another alternative, could be introducing a new method for that, something like:
``` ts
const fade: MotionAtom = {
  keyframes: [{ opacity: 0 }, { opacity: 1 }],
  duration: 500,
}

const scale: MotionAtom = {
  keyframes: [{ transform: "scale(0)" }, { transform: "scale(1)" }],
  duration: 800,
}

/* Single */
const Fade = createAtom(fade)
const Scale = createAtom(scale)

/* Multiple */
const FadeScale = createAtoms([scale, fade])
```

Please note that the `createAtoms` name is just to differentiate. It can be anything else that makes more sense.

Comment (layershifter) [reply rfcs/react-components/convergence/motion-definition-n-apis.md:328]: https://github.com/layershifter/office-ui-fabric-react/blob/docs/motion-rfc/rfcs/react-components/convergence/motion-definition-n-apis.md#motion-sequencing--grouping

@marcosmoura this is out of scope as I have not seen usages in our codebase or requests to implement this yet. It's not a big problem to implement, this have not been raised as a requirement yet.

**Review by ling1726 — APPROVED** [review]

**Review by marcosmoura — APPROVED** [review]

**Comment by layershifter** [comment]
A set of follow up issues is created:

- #30700
- #30699
- #30698
- #30697
- #30696
- #30695
